### DISCUSSION TITLE
what is behind modules, groovy and reflection

### DISCUSSION THREAD

**Discussion by soloturn** [MovingBlocks/Terasology#5110] [General]
setup development of terasology works with:
```
❯ ./groovyw module init iota
```

it downloads an ancient gradle version, uses a quite old jdk-11. what is the reasoning to do it that way, with groovy, reflection, and modules, instead of a "normal" java build?

**Comment by Cervator** [comment]
Hey there :-)

Define "normal" Java build? 😅 

Terasology has been around for quite a while and with sporadic efforts it is limited what gets upgraded. There's on-going work now to upgrade the Java version and Gradle.

The `groovyw` bit is some utility to make it easier to deal with cobbling together a workspace with a given set of modules, where "iota" is a basic starter collection. 

Reflection is just used in various ways based on whatever made sense at the time, really. Happy to hear of interest to improve things, especially if that comes with an offer to help! :-)

**Comment by skaldarnar** [comment]
Hi @soloturn, and welcome.

I just saw #5111 popping, pretty much the right time to look into Gradle and project setup, as we are trying to revamp it.

We've usually tried to stay up-to-date at least with the Gradle releases, but the project has gone through some quiet phase for a couple of months. @keturn started working on this (see https://github.com/MovingBlocks/Terasology/issues/4653) but went on hiatus soon after.
Currently @DarkWeird and @BenjaminAmos are picking this back up, see https://github.com/MovingBlocks/Terasology/pull/5109.

The "ancient Java" part is different though, as we cannot _just_ jump to a newer version. Some details can be found in https://github.com/MovingBlocks/Terasology/issues/3976 if you want to read up.

There are active discussion happening on our [Discord](https://discord.gg/terasology), we'd love to have you join the conversation, especially if you have some previous experience either with upgrading a code base to newer Gradle/Java.
